ReactNativeClient/lib/joplin-renderer/MdToHtml/rules/video.js

```diff
@@ -0,0 +1,70 @@
+// LOGIC: Extract the & UUID of the youtube video and the host name from the href link
+// and render it in iframe if both of the are valid.
+const url = require('url');
+const videoCss = () => {
+	return [
+		{
+			inline: true,
+			mime: 'text/css',
+			text: `
+				.jop-video-container {
+					position: relative;
+					width: 100%;
+					height: 0;
+					padding-bottom: 56.25%;
+				}
+
+				.jop-video-iframe {
+					position: absolute;
+					top: 0;
+					left: 0;
+					width: 100%;
+					height: 100%;
+				}`,
+		},
+	];
+};
+
+function addContextAssets(context) {
+	if ('video' in context.pluginAssets) return;
+
+	context.pluginAssets['video'] = videoCss();
+}
+
+function installRule(markdownIt, mdOptions, ruleOptions, context) {
+	const defaultRender = markdownIt.renderer.rules.link_open;
+
+	const videoIDRegex = /^.*(youtu\.be\/|v\/|u\/\w\/|embed\/|watch\?v=|&v=)([^#&?]*).*/;
+	const validHostName = ['youtu.be', 'www.youtube.com','www.youtube-nocookie.com'];
+	// videoIDRegex extracts the unique 11 digit code which is unique to every youtube video
+
+	markdownIt.renderer.rules.link_open = (tokens, idx, options, env, self) => {
+
+		const renderLink = defaultRender(tokens, idx, options, env, self);
+		const token = tokens[idx];
+		const link = token.attrGet('href');
+		const videoID = link.match(videoIDRegex);
+		const hostname = url.parse(link).hostname;
```

Comment:
If you use a url parser you no longer need your regex. Just print the result of the url.parse call and you'll see you have all the elements that you need.

Reply:
Sir since youtube supports a number of link format its preferable to just use regex, because if we use the passed URL method, we will end up checking for a lot of cases, let me explain
 
Youtube currently supports 4 types of link
|       |      |
| --- | --- |
|normal  http://www.youtube.com/watch?v=USjZcfj8yxE
shortened https://youtu.be/USjZcfj8yxE
no-cookie - http://www.youtube-nocookie.com/v/USjZcfj8yxE?version=3&hl=en_US&rel=0
legacy -

Reply:
Sorry but I don't want to go on and on about this URL parsing thing, this is a solved problem and there's already a parser in Joplin, so please use this.

I know you can use a regex, but yours had various issues and I personally can't make sense of it (as they said "he had a problem and tried to solve it with a regex, and now he has two problems). The parser does that job for you and has been battle tested in thousands of projects, so that's what we should use if we want something reliable.

Reply:
Alright, sir, I would implement this with URL parsing, but I seriously urge to read this,

the current regex had the bug you mentioned 
> https://www.a.com/watch?v=SADub7W22Zg => should not create a video player (it currently does!)

because the regex should only be used when you know that you are dealing with a youtube URL. Else it  caused false positives, e.g. this passes: 'v/2059-9080-5437'

So now that we do check that we are dealing with a youtube link it shouldn't have any problem.
you can check here
https://stackoverflow.com/a/8260383/12809166

Reply:
> Sorry but I don't want to go on and on about this URL parsing thing, this is a solved problem and there's already a parser in Joplin, so please use this.
> 
> I know you can use a regex, but yours had various issues and I personally can't make sense of it (as they said "he had a problem and tried to solve it with a regex, and now he has two problems"). The parser does that job for you and has been battle tested in thousands of projects, so that's what we should use if we want something reliable.

Sir I have made a change using the npm package [get-video-id](https://www.npmjs.com/package/get-video-id) used for extracting video ID for youtube and Vimeo, judging for the comment you made, you not keen on using this dependency and want to use the preexisting URL parser, but I seriously urge you to take a look at the changes.

Reply:
@laurent22  sir the extraction of youtube video ID is quite complex because you have to take into account for a variety of links, I tried to make a function which extracts the videoID but, it works fine for URLs which have video id in the query but URLs which don't have video id in querystring, then have to be extracted from path, and there you'll end up using some sort of regex to extract the video, I seriously believe the StackOverflow regex for getting videoID will work for 99% of the cases.
https://repl.it/@Rishgod/VideoID-extraction
_(I tried to implement it using parsing url)_